src/matrixrtc/MatrixRTCSessonManager.ts

```diff
@@ -0,0 +1,130 @@
+        // returing nothing, and breaks tests if you change it to return an empty array :'(
+        for (const room of this.client.getRooms() ?? []) {
+            const session = MatrixRTCSession.roomSessionForRoom(this.client, room);
+            if (session.memberships.length > 0) {
+                this.roomSessions.set(room.roomId, session);
+            }
+        }
+
+        this.client.on(ClientEvent.Room, this.onRoom);
+        this.client.on(RoomStateEvent.Events, this.onRoomState);
+    }
+
+    public stop(): void {
+        for (const sess of this.roomSessions.values()) {
+            sess.stop();
+        }
+        this.roomSessions.clear();
+
+        this.client.removeListener(ClientEvent.Room, this.onRoom);
+        this.client.removeListener(RoomStateEvent.Events, this.onRoomState);
+    }
+
+    /**
+     * Get a list of all ongoing MatrixRTC sessions that have 1 or more active
+     * members
+     * (whether the client is joined to them or not)
+     */
+    public getActiveSessions(): MatrixRTCSession[] {
+        return Array.from(this.roomSessions.values()).filter((m) => m.memberships.length > 0);
+    }
+
+    /**
+     * Gets the main MatrixRTC session for a room, or undefined if there is
+     * no current session
+     */
+    public getActiveRoomSession(room: Room): MatrixRTCSession | undefined {
+        return this.roomSessions.get(room.roomId)!;
+    }
+
+    /**
+     * Gets the main MatrixRTC session for a room, returning an empty session
+     * if no members are currently participating
+     */
+    public getRoomSession(room: Room): MatrixRTCSession {
+        if (!this.roomSessions.has(room.roomId)) {
+            this.roomSessions.set(room.roomId, MatrixRTCSession.roomSessionForRoom(this.client, room));
+        }
+
+        return this.roomSessions.get(room.roomId)!;
+    }
+
+    private onRoom = (room: Room): void => {
+        this.refreshRoom(room);
+    };
+
+    private onRoomState = (event: MatrixEvent, _state: RoomState): void => {
+        const room = this.client.getRoom(event.getRoomId());
+        if (!room) {
+            logger.error(`Got room state event for unknown room ${event.getRoomId()}!`);
+            return;
+        }
+
+        this.refreshRoom(room);
+    };
+
+    private refreshRoom(room: Room): void {
+        const hadSession = this.roomSessions.has(room.roomId);
+        const sess = this.getRoomSession(room);
+
+        const wasActive = sess.memberships.length > 0 && hadSession;
+
+        sess.onMembershipUpdate();
+
+        const nowActive = sess.memberships.length > 0;
+
+        if (wasActive && !nowActive) {
+            this.emit(MatrixRTCSessionManagerEvents.SessionEnded, room.roomId, this.roomSessions.get(room.roomId)!);
+        } else if (!wasActive && nowActive) {
+            this.emit(MatrixRTCSessionManagerEvents.SessionStarted, room.roomId, this.roomSessions.get(room.roomId)!);
+        }
+    }
```

Comment:
This function is the only place where we emit events about sessions being started and stopped from what I can see, so I assume that we must ensure that it always properly emits events.

I have a couple of notes regarding the topic of `hadSession`, let me know if they make sense to you:
1. `hadSession` seems to be `true` when the session is known to the session manager at the moment of calling this function and `false` otherwise. This means that `wasActive` will be `false` even if the session _was active_ unless the session was known to the session manager which makes the name `wasActive` a bit confusing. I think it's probably easier to split `wasActive` and `hadSession` (perhaps renaming it to `knownSession` or calling it `newSession` which will be an inverse of the current `hadSession`), otherwise `wasActive` and `nowActive` don't look symmetrical.
2. Given that `this.roomSessions` only grows during the lifetime of a session manager, won't it introduce some side effects in this function given that `hadSession` is there to specifically treat the case of session being known / not known to the session manager? (i.e. stopped sessions/sessions) with 0 members will remain on the map.

To me it feels like the most elegant to solve it would be to make sure we don't store expired sessions in the map (this may eliminate the need of the `hadSession`).

Reply:
Yeah, this is a little confusing because `this.getRoomSession()` will create the session if it doesn't exist, and from the current room state, so sessions that were not previously in the map would be created reflecting the update that just happened. We could use `getActiveRoomSession()` instead here which would return null if we didn't know about it, but we need to create a new one anyway in order to tell it to update. I've renamed the variables, perhaps this makes it a bit clearer? I agree it's not the world's most elegant code though.

The problem with removing stopped sessions from the map is that only the sessions in this map will get updates dispatched to them, so if anything holds on to a session object that we've removed from the map, it will no longer reflect the current session in that room because that would then be a new object. Keeping the sessions around ensures that we always only ever use one object per room session for the lifetime of the client, so no possibility of stale objects. I'll add a comment to this effect.

Reply:
Agreed that adding something to get a call from the Room object could be quite nice, as long as it is just a glue method and the actual code is somewhere else because rooms are already quite overloaded (room.ts is about 3k lines!). I mostly did it this was because this was similar to how the previous GroupCall stuf worked. I'm not sure how the session manager could be bottleneck though: it's not like it's running on its own thread/coroutine.

I was envisaging separate accessors for room session and user session to make it as easy as possible to get a room session, and we only ever support one of these per room by definition. I would add a separate `getUserSessions(): MatrixRTCSessiion[]` once we support user sessions (and therefore multiple sessions per room).

I'm not quite sure I understand your `MatrixRTCSession ` proposal - what are the `Member`/`LocalMember` objects / what do they represent exactly? 

Either way, this is sounding like a relatively significant refactor, so one option here could be to revisit this when we add support for user sessions?

Reply:
> I'm not quite sure I understand your MatrixRTCSession proposal - what are the Member/LocalMember objects / what do they represent exactly?

Members/participants of a call. I.e. similarly to `Participant`, `RemoteParticipant`, and `LocalParticipant` in a LiveKit. Each `Participant` holds some data (room membership data), `LocalParticipant` would then be a type that on top of the regular data that each `Participant` holds, has some additional methods, that e.g. allow changing participant's metadata. Changing it would consequently trigger a `MemberChanged` event so that other participants can react to the changes on their side.

> Either way, this is sounding like a relatively significant refactor, so one option here could be to revisit this when we add support for user sessions?

Yeah, we can do it like this I think.